presto-main/src/test/java/com/facebook/presto/operator/AnnotationBasedOperators.java

```diff
@@ -0,0 +1,696 @@
+            return v;
+        }
+    }
+
+    public static class StaticMethodScalarFunction
+    {
+        @ScalarFunction("static_method_scalar")
+        @Description("Simple scalar with single implemenatation based on method")
+        @SqlType(DOUBLE)
+        public static double fun(@SqlType(DOUBLE) double v)
+        {
+            return v;
+        }
+    }
+
+    public static class MultiScalarFunction
+    {
+        @ScalarFunction("static_method_scalar_1")
+        @Description("Simple scalar with single implemenatation based on method 1")
+        @SqlType(DOUBLE)
+        public static double fun1(@SqlType(DOUBLE) double v)
+        {
+            return v;
+        }
+
+        @ScalarFunction(value = "static_method_scalar_2", hidden = true, deterministic = false)
+        @Description("Simple scalar with single implemenatation based on method 2")
+        @SqlType(BIGINT)
+        public static long fun2(@SqlType(BIGINT) long v)
+        {
+            return v;
+        }
+    }
+
+    @ScalarFunction("parametric_scalar")
+    @Description("Parametric scalar description")
+    public static class ParametricScalarFunction
+    {
+        @SqlType("T")
+        @TypeParameter("T")
+        public static double fun(@SqlType("T") double v)
+        {
+            return v;
+        }
+
+        @SqlType("T")
+        @TypeParameter("T")
+        public static long fun(@SqlType("T") long v)
+        {
+            return v;
+        }
+    }
+
+    @ScalarFunction("with_exact_scalar")
+    @Description("Parametric scalar with exact and generic implementations")
+    public static class ComplexParametricScalarFunction
+    {
+        @SqlType(BOOLEAN)
+        @LiteralParameters("x")
+        public static boolean fun1(@SqlType("array(varchar(x))") Block array)
+        {
+            return true;
+        }
+
+        @SqlType(BOOLEAN)
+        public static boolean fun2(@SqlType("array(varchar(17))") Block array)
+        {
+            return true;
+        }
+    }
+
+    @ScalarFunction("parametric_scalar_inject")
+    @Description("Parametric scalar with literal injected")
+    public static class SimpleInjectionScalarFunction
+    {
+        @SqlType(BIGINT)
+        @LiteralParameters("x")
+        public static long fun(@LiteralParameter("x") Long literalParam, @SqlType("varchar(x)") Slice val)
+        {
+            return literalParam;
+        }
+    }
+
+    @ScalarFunction("parametric_scalar_inject_constructor")
+    @Description("Parametric scalar with type injected though constructor")
+    public static class ConstructorInjectionScalarFunction
+    {
+        private final Type type;
+
+        @TypeParameter("T")
+        public ConstructorInjectionScalarFunction(@TypeParameter("T") Type type)
+        {
+            this.type = type;
+        }
+
+        @SqlType(BIGINT)
+        @TypeParameter("T")
+        public long fun(@SqlType("array(T)") Slice val)
+        {
+            return 17L;
+        }
+    }
```

Comment:
Add test for cases where the injected SqlType is `map(varchar, int)` or `map(K, V)`. Your current test covers injecting `T`, it doesn't cover injecting concrete types, or types that contains (but is not only) T.

This feature is recently added primarily by relaxing sanity checks. It was added after your pull request was put up. I'm sorry to have you incorporate it into your pull request.